File: Models/Task.swift
//
//  Task.swift
//  TickTick
//
//  Task model
//

import Foundation

struct Task: Identifiable, Codable, Equatable {
    var id: UUID
    var title: String
    var notes: String
    var isCompleted: Bool
    var dueDate: Date?
    var priority: Priority
    var listId: UUID
    var createdAt: Date
    var completedAt: Date?
    
    init(
        id: UUID = UUID(),
        title: String,
        notes: String = "",
        isCompleted: Bool = false,
        dueDate: Date? = nil,
        priority: Priority = .none,
        listId: UUID,
        createdAt: Date = Date(),
        completedAt: Date? = nil
    ) {
        self.id = id
        self.title = title
        self.notes = notes
        self.isCompleted = isCompleted
        self.dueDate = dueDate
        self.priority = priority
        self.listId = listId
        self.createdAt = createdAt
        self.completedAt = completedAt
    }
}

enum Priority: String, Codable, CaseIterable {
    case none = "None"
    case low = "Low"
    case medium = "Medium"
    case high = "High"
    
    var color: String {
        switch self {
        case .none: return "gray"
        case .low: return "blue"
        case .medium: return "orange"
        case .high: return "red"
        }
    }
}

struct TaskList: Identifiable, Codable {
    var id: UUID
    var name: String
    var color: String
    var icon: String
    
    init(id: UUID = UUID(), name: String, color: String = "blue", icon: String = "list.bullet") {
        self.id = id
        self.name = name
        self.color = color
        self.icon = icon
    }
}

File: web-app/src/types/Task.ts
export interface Task {
  id: string
  title: string
  notes: string
  isCompleted: boolean
  dueDate?: Date
  priority: Priority
  listId: string
  createdAt: Date
  completedAt?: Date
  tags?: string[]
  subtasks?: Subtask[]
  recurring?: RecurringPattern
  reminderDate?: Date
  parentTaskId?: string
  attachmentIds?: string[]
  dependsOn?: string[] // IDs of tasks this task depends on
  blockedBy?: string[] // IDs of tasks that depend on this task
  milestoneId?: string // Linked milestone
  assigneeId?: string // Assigned user ID
  customFieldValues?: Record<string, any> // Custom field values
  approvalId?: string // Linked approval
  okrKeyResultId?: string // Linked OKR key result
}

export interface Subtask {
  id: string
  title: string
  isCompleted: boolean
  createdAt: Date
  completedAt?: Date
}

export interface RecurringPattern {
  type: 'daily' | 'weekly' | 'monthly' | 'yearly' | 'custom'
  interval: number
  endDate?: Date
  daysOfWeek?: number[] // 0 = Sunday, 1 = Monday, etc.
  dayOfMonth?: number
}

export enum Priority {
  None = 'none',
  Low = 'low',
  Medium = 'medium',
  High = 'high'
}

export interface TaskList {
  id: string
  name: string
  color: string
  icon: string
  backgroundImage?: string
  backgroundColor?: string
}

export const PriorityColors: Record<Priority, string> = {
  [Priority.None]: '#8E8E93',
  [Priority.Low]: '#007AFF',
  [Priority.Medium]: '#FF9500',
  [Priority.High]: '#FF3B30'
}

export const PriorityLabels: Record<Priority, string> = {
  [Priority.None]: 'None',
  [Priority.Low]: 'Low',
  [Priority.Medium]: 'Medium',
  [Priority.High]: 'High'
}

